fs: add support for DirBuilder and DirBuilderExt

shkurski opened the pull request:
## Motivation

`tokio` doesn't have an asynchronous version of `std::fs::DirBuilder` yet.
This PR adds support for `std::fs::DirBuilder` and implementation for the `std::os::unix::fs::DirBuilderExt` trait.

## Solution

The initial idea was to implement  a thin wrapper  around an internally
held `std::fs::DirBuilder` instance.  This, however, didn't work due to
`std::fs::DirBuilder` not having Copy/Clone traits implemented, which
was necessary for constructing an instance to move-capture it into a
closure.

Instead,  we have mirrored the `std::fs::DirBuilder` configuration by  storing
the `recursive` and (unix-only) `mode`  parameters locally,  which were then
used to construct the `std::fs::DirBuilder` instance on-the-fly.

The (unix-only) `DirBuilderExt` trait  has  been  implemented  in-place
within the `dir_builder` crate. Despite it might have looked better to
have it separated somewhere in `../fs/os/unix`, I haven't found a clean
and idiomatic way to do so (due to Rust's "orphan rules").  I'd be glad
to hear  the propositions  in case someone  sees  a  way for this to be
further simplified.

Edit: force-pushed immediately after the initial commit was made.
Hope it was quick enough to not break any local repo's.

Fixes: #2369

Darksonn commented on tokio/src/fs/dir_builder.rs:
````diff
@@ -0,0 +1,137 @@
+    /// let builder = DirBuilder::new();
+    /// ```
+    pub fn new() -> DirBuilder {
+        #[cfg(not(unix))]
+        let builder = DirBuilder { recursive: false };
+
+        #[cfg(unix)]
+        let builder = DirBuilder {
````
We can just use the `Default` impl here.
```rust
pub fn new() -> Self {
    Default::default()
}
```

shkurski replied:
Thank you, [switched](https://github.com/tokio-rs/tokio/pull/2524/commits/384d1850e74d529666871a7008e44724b9ea6a3b) to the `Default` impl.

Darksonn commented on tokio/src/fs/dir_builder.rs:
```diff
@@ -0,0 +1,137 @@
+        let mut builder = std::fs::DirBuilder::new();
+        builder.recursive(self.recursive);
+
+        #[cfg(unix)]
+        {
+            if let Some(mode) = self.mode {
```
This can go directly on the if.

Darksonn commented on tokio/src/macros/cfg.rs:
```diff
@@ -354,6 +354,15 @@ macro_rules! cfg_uds {
     }
 }
 
+macro_rules! cfg_unix {
+    ($($item:item)*) => {
+        $(
+            #[cfg(unix)]
+            $item
+        )*
+    }
+}
```
We use `#[cfg(unix)]` everywhere else in the codebase. Maybe adding this `cfg_*` makes sense, but I would rather want it in its own PR.

shkurski replied:
[Removed](https://github.com/tokio-rs/tokio/pull/2524/commits/384d1850e74d529666871a7008e44724b9ea6a3b).

Darksonn commented on tokio/src/fs/dir_builder.rs:
```diff
@@ -0,0 +1,137 @@
+/// Additional Unix-specific options are available via importing the
+/// [os::unix::fs::DirBuilderExt] trait.
+///
+/// This is a specialized version of [`std::fs::DirBuilder`] for usage from
```
```suggestion
/// This is a specialized version of [`std::fs::DirBuilder`] for usage on
```

shkurski replied:
Thanks! [Fixed](https://github.com/tokio-rs/tokio/pull/2524/commits/e73a9b275af9cee1acb9e3466912e31e1e979646)

shkurski commented on tokio/src/fs/dir_builder.rs:
```diff
@@ -0,0 +1,137 @@
+/// the Tokio runtime.
+///
+/// [os::unix::fs::DirBuilderExt]: ../os/unix/fs/trait.DirBuilderExt.html
+/// [std::fs::DirBuilder]: https://doc.rust-lang.org/std/fs/struct.DirBuilder.html
```
[Fixed](https://github.com/tokio-rs/tokio/pull/2524/commits/e73a9b275af9cee1acb9e3466912e31e1e979646).

shkurski commented on tokio/src/fs/dir_builder.rs:
```diff
@@ -0,0 +1,137 @@
+/// A builder for creating directories in various manners.
+///
+/// Additional Unix-specific options are available via importing the
+/// [os::unix::fs::DirBuilderExt] trait.
```
[Fixed](https://github.com/tokio-rs/tokio/pull/2524/commits/e73a9b275af9cee1acb9e3466912e31e1e979646).

shkurski commented:
> For the same reason as described in [#2515](https://github.com/tokio-rs/tokio/pull/2515#issuecomment-626640636), we should define our own extension trait instead of using the one from std.

Thanks for the review :+1: 
I've implemented our own [extension trait](https://github.com/tokio-rs/tokio/pull/2524/commits/1f64f39d3a87570effeccd6121530cbe60f4a734) and addressed your comments.

Darksonn commented on tokio/src/fs/os/unix/dir_builder_ext.rs:
```diff
@@ -0,0 +1,29 @@
+use crate::fs::dir_builder::DirBuilder;
+
+/// Unix-specific extensions to [`DirBuilderExt`].
+///
+/// [crate::fs::os::unix::DirBuilderExt]: DirBuilderExt
```
This link seems off. Shouldn't it link to `DirBuilder`?
